In [131]:
import random

In [132]:
def c_scan(requests, head, disk_size):
    current_position = head
    seek_count = 0
#     seek_sequence = []
    worst_case_seek_time = 0
    distance = 0

    left = [0]
    right = [disk_size - 1]

    for req in requests:
        if req < head:
            left.append(req)
        elif req > head:
            right.append(req)

    left.sort()
    right.sort()

    for track in right:
#         seek_sequence.append(track)
        distance = abs(track - current_position)
        seek_count += distance
        worst_case_seek_time = max(worst_case_seek_time, distance)
        current_position = track
    
    
    current_position = 0
    distance = abs((disk_size - 1) - current_position)
    seek_count += distance
    worst_case_seek_time = max(worst_case_seek_time, distance)

    for track in left:
#         seek_sequence.append(track)
        distance = abs(track - current_position)
        seek_count += distance
        worst_case_seek_time = max(worst_case_seek_time, distance)
        current_position = track

#     print("Total number of seek operations =", seek_count)
#     print("Worst-case Seek Time =", worst_case_seek_time)
#     print("Seek Sequence is", *seek_sequence)

    avg_seek_time = seek_count / len(requests)
#     avg_seek_time = seek_count

#     print("Average Seek Time =", avg_seek_time)

 # Returning both average and worst-case seek times
    return avg_seek_time, worst_case_seek_time

In [133]:
def c_look(requests, head):
    """
    C-LOOK disk scheduling algorithm.
    :param requests: List of cylinder requests
    :param head: Current head position
    :return: Total seek count
    """
    size = len(requests)
    distance = 0
    cur_track = head

    # Sort the request list
    requests.sort()

    # Start processing the requests
    for i in range(size):
        next_track = requests[i]
        # Calculate the distance
        distance += abs(cur_track - next_track)
        # Update the current head position
        cur_track = next_track

    # Jumping back to the first request in the direction
    if size != 0:
        distance += abs(requests[0] - cur_track)
        cur_track = requests[0]

    return distance

In [134]:
# Initialize variables
disk_size = 200
initial_head_position = 50
num_simulations = 1000

In [135]:
# Different sets of requests
request_sizes = [10, 20, 50, 100]

In [136]:
# Dictionary to store results
results_c_scan = {}
results_c_look = {}

In [137]:
# def run_algorithm_simulation(algorithm, requests, request_sizes, initial_head_position, disk_size, num_simulations):
#     results = {}

#     for size in request_sizes:
#         total_seek = 0
#         worst_seek = 0

#         for _ in range(num_simulations):
# #             requests = random.sample(range(disk_size), size)
#             seek = algorithm(requests, initial_head_position, disk_size)
#             total_seek += seek
#             worst_seek = max(worst_seek, seek)

#         avg_seek = total_seek / num_simulations
#         results[size] = (avg_seek, worst_seek)

#     return results

In [138]:
# # Run C-LOOK algorithm for different sizes of requests
# for size in request_sizes:
#     total_seek = 0
#     worst_seek = 0
#     for _ in range(1000): # Run 1000 simulations for each set of requests
#         requests = random.sample(range(disk_size), size)
#         seek = c_look(requests, initial_head_position)
#         total_seek += seek
#         worst_seek = max(worst_seek, seek)

#     # Calculate average and worst-case seek times
#     avg_seek = total_seek / 1000
#     results[size] = (avg_seek, worst_seek)

In [139]:
# Run C-LOOK algorithm for different sizes of requests
for size in request_sizes:
    
    total_seek_c_scan = 0
    worst_seek_c_scan = 0
    
    total_seek_c_look = 0
    worst_seek_c_look = 0
    for _ in range(1000): # Run 1000 simulations for each set of requests
        requests = random.sample(range(disk_size), size)
        
        seek_c_scan,worst_seek_c_scan_temp = c_scan(requests,initial_head_position,disk_size)
        total_seek_c_scan += seek_c_scan
        worst_seek_c_scan = max(worst_seek_c_scan, worst_seek_c_scan_temp)
        
        seek_c_look = c_look(requests, initial_head_position)
        total_seek_c_look += seek_c_look
        worst_seek_c_look = max(worst_seek_c_look, seek_c_look)

    # Calculate average and worst-case seek times
    avg_seek_c_scan = total_seek_c_scan / 1000
    results_c_scan[size] = (avg_seek_c_scan,worst_seek_c_scan)
    
    avg_seek_c_look = total_seek_c_look / 1000
    results_c_look[size] = (avg_seek_c_look, worst_seek_c_look)

In [140]:
results_c_scan


{10: (38.00750000000001, 199),
 20: (19.408349999999984, 199),
 50: (7.8816199999999546, 199),
 100: (3.9597099999999577, 199)}

In [141]:
results_c_look

{10: (360.538, 448),
 20: (405.891, 448),
 50: (432.802, 448),
 100: (442.946, 448)}

Analysis

The average seek time increases with the number of requests. This is expected as more requests generally mean more movement for the disk head.

The worst-case seek time remains constant at 448 cylinders for all sets of requests. This is because the worst-case scenario in C-LOOK is determined by the furthest request from the current head position, and since the disk size and head position are constant, the maximum possible seek distance doesn't change.

The C-LOOK algorithm performs efficiently in terms of reducing the total back-and-forth movement of the disk head, as seen by the relatively low average seek times even for larger sets of requests. This makes it suitable for systems where minimizing the disk arm movement is crucial for performance.